File: 948-bag-of-tokens/948-bag-of-tokens.js
const countPoints = (tokens, P) => {
  let [i, score, sum] = [0, 0, 0];
  for (i; sum + tokens[i] <= P; score++, i++) {
    sum += tokens[i];
  }
  return score;
}

const bagOfTokensScore = (tokens, P, max = 0) => {
  const len = tokens.length;
  if (max === 0) tokens.sort((a, b) => a - b);
  if (len === 0) return max;
  const score = countPoints(tokens, P);
  if (score > max) max = score;
  return P > tokens[0]
    ? bagOfTokensScore(tokens.slice(1, len - 1), P - tokens[0] + tokens[len - 1], max)
    : max;
};

File: 948-bag-of-tokens/948-bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(),tokens.end());
        int n=tokens.size();
        if(n==0)
            return 0;
        if(power<tokens[0])
            return 0;
        int currsum=0;
        int maxsum=0;
        int last=n-1;
        for(int i=0;i<=last;i++)
        {
            maxsum=max(maxsum,currsum);
            if(power>=tokens[i])
            {
                currsum++;
                power-=tokens[i];
            }
            else
            {
                power+=tokens[last];
                currsum--;
                last--;
                
                if(power>=tokens[i])
                {
                    currsum++;
                    power-=tokens[i];
                }
                else
                {
                    break;
                }
            }
        }
        maxsum=max(maxsum,currsum);
        return maxsum;
    }
};